src/sources/http_scrape/mod.rs

```diff
@@ -0,0 +1,212 @@
+
+use bytes::Bytes;
+use futures_util::{stream, FutureExt, StreamExt, TryFutureExt};
+use http::{response::Parts, Uri};
+use hyper::{Body, Request};
+use std::time::{Duration, Instant};
+use std::{collections::HashMap, future::ready};
+use tokio_stream::wrappers::IntervalStream;
+
+use crate::{
+    http::{Auth, HttpClient},
+    internal_events::{
+        EndpointBytesReceived, HttpScrapeEventsReceived, HttpScrapeHttpError,
+        HttpScrapeHttpResponseError, RequestCompleted, StreamClosedError,
+    },
+    tls::TlsSettings,
+    Error, SourceSender,
+};
+use vector_common::shutdown::ShutdownSignal;
+use vector_core::{config::proxy::ProxyConfig, event::Event, ByteSizeOf};
+
+/// Contains the inputs generic to any http scrape.
+pub(crate) struct GenericHttpScrapeInputs {
+    /// Array of URLs to scrape
+    pub urls: Vec<Uri>,
+    /// Interval to scrape on in seconds
+    pub interval_secs: u64,
+    /// Map of Header+Value to apply to HTTP request
+    pub headers: Option<HashMap<String, Vec<String>>>,
+    /// Content type of the HTTP request, determined by the source
+    pub content_type: String,
+    pub auth: Option<Auth>,
+    pub tls: TlsSettings,
+    pub proxy: ProxyConfig,
+    pub shutdown: ShutdownSignal,
+}
+
+/// The default interval to scrape the http endpoint if none is configured.
+pub(crate) const fn default_scrape_interval_secs() -> u64 {
+    15
+}
+
+/// Methods that allow context-specific behavior during the scraping procedure.
+pub(crate) trait HttpScraper {
+    /// (Optional) Called before the HTTP request is made, allows building context.
+    fn build(&mut self, _url: &Uri) {}
+
+    /// Called after the HTTP request succeeds and returns the decoded/parsed Event array.
+    fn on_response(&mut self, url: &Uri, header: &Parts, body: &Bytes) -> Option<Vec<Event>>;
+
+    /// (Optional) Called if the HTTP response is not 200 ('OK').
+    fn on_http_response_error(&self, _uri: &Uri, _header: &Parts) {}
+}
+
+/// Builds a url for the HTTP requests.
+pub(crate) fn get_url(uri: &Uri, query: &Option<HashMap<String, Vec<String>>>) -> Uri {
+    let mut serializer = url::form_urlencoded::Serializer::new(String::new());
+    if let Some(query) = uri.query() {
+        serializer.extend_pairs(url::form_urlencoded::parse(query.as_bytes()));
+    };
+    if let Some(query) = &query {
+        for (k, l) in query {
+            for v in l {
+                serializer.append_pair(k, v);
+            }
+        }
+    };
+    let mut builder = Uri::builder();
+    if let Some(scheme) = uri.scheme() {
+        builder = builder.scheme(scheme.clone());
+    };
+    if let Some(authority) = uri.authority() {
+        builder = builder.authority(authority.clone());
+    };
+    builder = builder.path_and_query(match serializer.finish() {
+        query if !query.is_empty() => format!("{}?{}", uri.path(), query),
+        _ => uri.path().to_string(),
+    });
+    builder.build().expect("error building URI")
```

Comment:
I suggest making errors like this more descriptive, particularly if you can say why it should never fail (and if it can legitimately fail, don't use `expect`/`unwrap`). For example, `expect("Failed to rebuild URI from parsed parts")`

Reply:
👍  I can try to improve on that. These bits were direct from the `prometheus_scrape` code, but always a good opportunity to improve.

Reply:
Yeah, copied code shows up as if it were new, prompting us to re-review what you didn't actually write.